### DISCUSSION TITLE
Faster builds/rebuilds for development and production ⚡️

### DISCUSSION THREAD

**Discussion by renanccastro** [meteor/meteor#11587] [Feature Requests]
**Hello, everyone!**

It's well known that for some time our meteor builds can take longer than the average players in the JS world, like webpack and rollup, for example, but, we also do more! We have nested imports, dynamic imports, and custom package resolution, everything without a single config, which is something you won't find out there :) We have made a lot of progress in this matter lately, which is great! First-time builds and bundling however are still slow for some apps.

Babel often takes up so much time (50%+) during initial (cold start) meteor builds (by @benjamn , as in this answer: https://github.com/meteor/meteor-feature-requests/issues/398#issuecomment-856081279) that it would be a good thing to replace it if we find a better alternative.

This is the profiling result for a bigger app, we can see that Babel is indeed 50% of the initial build. Rebuilds are faster.

```
| Top leaves:
| Babel.compile...........................................36,901 ms (1180)
| files.stat...............................................4,573 ms (67490)
| other ImportScanner#findImportedModuleIdentifiers........3,487 ms (2918)
| files.readFile...........................................1,726 ms (15431)
| sha1.....................................................1,543 ms (13988)
| other reifyCompileWithCache..............................1,260 ms (1899)
| other ImportScanner#resolve................................956 ms (13534)
| linker File#getPrelinkedOutput.............................817 ms (4136)
| getPrelinkedFiles toStringWithSourceMap....................798 ms (196)
| other ImportScanner#scanImports for the app................603 ms (2)
| files.watchFile............................................563 ms (91718)
| linker File#computeAssignedVariables.......................549 ms (3251)
| files.writeFile............................................482 ms (5116)
| other safeWatcher.watch....................................415 ms (88359)
| other PackageSource#_findSources for os....................399 ms (1)
| other linker Module#getPrelinkedFiles......................284 ms (196)
| other ImportScanner#scanImports for base64.................224 ms (2)
| other bundler writeSiteArchive.............................211 ms (1)
| other ImportScanner#scanImports for tmeasday:check-npm-versions.204 ms (2)
| other ImportScanner#realPath...............................193 ms (5067)
| other compileUnibuild (the app)............................183 ms (4)
| other JsImage#write........................................182 ms (1)
| files.readdirWithTypes.....................................151 ms (18954)
| other PackageSource#_findSources for web.browser...........135 ms (1)
| files.lstat................................................122 ms (17183)
| other ClientTarget#write...................................108 ms (1)
| other optimistic readJsonOrNull............................105 ms (4858)
| other Target#_emitResources................................104 ms (2)
| other Builder#copyNodeModulesDirectory.....................103 ms (36)
| 
| (#3) Total: 60,382 ms (Build App)
```
So, let's discuss ideas for improving build/rebuild time on Meteor!
Here are some already pitched ideas, and some status:

## Use SWC as a replacement for Babel

Initial feature request surfaced from ben's answer here: https://github.com/meteor/meteor-feature-requests/issues/398#issuecomment-856081279
**Status:** Tested, there are multiple blockers for using it.

The general idea is that SWC uses a rust backend and, on benchmarks, it's way faster than Babel.
What I've tested so far is getting a build plugin that uses SWC as a backend but there are some key points that block meteor from using it:

1. Documentation is mainly for Rust, and the JS side is not up-to-date with its rust backend (Ex is the babel-compatible AST, which is possible on rust, but not yet on JS )
2. https://github.com/swc-project/swc/issues/1392 - it's not possible currently to generate babel-compatible AST, which means that Reify, which plugs inside ecmascript package in a babel plugin, for parsing our modules can't run without a rewrite. SWC ast could probably be made compatible with reify with some work (I've advanced at some point that a simple program would work), but that leaves us to the next point
3. It doesn't support non-top-level imports when parsing: https://github.com/swc-project/swc/blob/7488950f90bba8c454a12a611d1f8d77405feac7/ecmascript/parser/src/parser/stmt/module_item.rs#L632 and that is rust code

It's possible to progress more on this matter, and if anyone is interested, check out my POC here: https://github.com/renanccastro/swc-experiment

## Use esbuild as an alternative for our Bundler

Source feature request: https://github.com/meteor/meteor-feature-requests/issues/398
**Status:** Pending

## Use Multiple cores on our own bundling/builder 

**Status:** Pending
Worker threads would be a good fit for tree resolution.

## Use Webpack 

**Status:** Pending

There is already a working package that let you use webpack inside Meteor: https://github.com/ardatan/meteor-webpack
We need to analyze and benchmark the performance and its tradeoffs.

--------
Also, feel free to suggest new ideas.

**Comment by santiagopuentep** [comment]
What would be the difficulty of instead of patching the system, change the paradigm completely and move to a completely different one like Vite or Snowpack (which don't bundle, they use the native ES modules system and change only the affected file)? I think that we could spend a lot of time improving bundling performance, but the future seems to be bundle-less.

It´s such a huge pain to wait for Meteor on every change, especially on Windows. It's the only big complain around Meteor.

**Reply by renanccastro** [reply]
I like this direction too @santiagopuentep. It's not exactly an easy path, but I think it might be the best one.
We would have to change our way to load modules on the client-side, which would also change our way to do dynamic imports. 

Anyway, I think the first task for this would be to list all the points we will have to change and post them in public so we can discuss alternatives.

Thanks for your feedback here

**Reply by santiagopuentep** [reply]
Yeah, without a change like this one, I don't think Meteor can ever become a big scale thing. Large projects are too slow, and this will get worse in the future IMO. Our project for example would have to be moved to some other system if the build times keep getting worse and worse. It's a shame because Meteor is an amazing platform otherwise.

With a lightning fast build system and moving everything to independent npm modules, I think this could be a future proof platform.

**Reply by brucejo75** [reply]
[isobuild](https://github.com/meteor/meteor/tree/devel/tools/isobuild) has long been a differentiator for Meteor.  But I think the progress of `ecsmascript` combined with solutions like [Vite](https://vitejs.dev/) have eroded that advantage.

I am under the impression that [isobuild](https://github.com/meteor/meteor/tree/devel/tools/isobuild) has required a lot of attention over the last 3-4 years.  Being able to leverage another tool may be a big productivity win.

But... I know the cost to transition would likely be high.

Tough spot.

**Comment by paulincai** [comment]
To be frank, we could start by defining some terms like "large project", "my Windows machine".
I never though Meteor was slow until I red this discussion and it makes me feel like I am missing a point of reference or I use a wrong paradigm.
I am curious how exactly is productivity affected in your case. Are you referring to dev builds or production? Are you deploying very often? Do you use CI? In general CI might be a bottleneck and not the local machine.
For development in particular, everything I change reflects instantly in my browser. Are things any different for you?

Regardless of the time it takes to build a production bundle, I think we perceive slowness in our day to day development habits and if this is the case, I am not sure the tech needs a change though I am not against progress.

Some of the things I do to prevent having to  build a large project:
- microservices
- microfrontend - ing
- keeping an up to date modern-browser configuration to prevent building legacy JS into the modern bundle.
- reading through the code of NPMs  I use and fork when necessary to make sure I don't add "noise" or junk to my projects.

Some other things I do to keep existing projects small:
- use my own code where an NPM does a lot more than I need
- use "lodash.function" NPMs instead of getting the whole lodash.
- for players(e.g HLS), maps etc that have a CDN option, I add them as HTML head tags as I need them, instead of importing them in my bundle. For other libraries that I can easily convert to a small js file and deliver it through my own CDN, same as above.

One of my top priority rules in a project is that I should not have a large project bundle. It is best to run multiple functional components of a larger project. With this in mind, I would rather focus on an official SSO and an official microfrontend-ing technology to attach Meteor and non-Meteor apps to one another in a "master app".

**Reply by santiagopuentep** [reply]
In Windows here Meteor is almost unusable. I'm using an Core i3 right now, but with 16 GB of RAM. Everything works just fine, like Storybook for example. I run Meteor and I get 100% CPU usage. In faster computers we get 30% or 40% usage, just by running Meteor. On mac the reload times very fast, in Windows they take forever.

The platform that doesn't strive to be the best, doesn't survive long term in the JS world. Having a good enough bundler is not enough, Vite doesn't even bundle in development. How can you compete with that? Meteor looks like an old dinosaur already, I love Meteor, but it's true. If it continues like this it won't survive a new technology doing something similar, but with Vite, full NPM support, etc.

**Reply by afrokick** [reply]
@paulincai thanks for some tips.

We decided to split our large app (500k LOC) into microfrontoends(MC) and build it with ESBuild.

In two words:
We have a core(meteor app), which contain all common code like ui library, state manager and some business logic + meteor packages.

Every MC is npm package.

It is building with ESBuild. The config for ESBuild looks like:

```
import { exec } from 'child_process';
import path from 'path';
import { build } from 'esbuild';
import { lessLoader } from 'esbuild-plugin-less'; // as example, we use LESS for styles.

// isProduction flag for watch mode
const isProduction = process.env.NODE_ENV === 'production';
const MODULE_NAME = process.env.npm_package_name ?? '';
const NAME_PARTS = MODULE_NAME.split('/');
const isPrefixed = NAME_PARTS.length > 1;

const SAFE_NAME = NAME_PARTS.join('_'); // convert @company/package1 -> @company_package1

const DEV_LIVE_MODULES_PATH = path.resolve(`../../../public/dev-live-modules/`); //adjust it to your project structure
const OUTPUT_FILE_PATH = path.resolve(`./dist/${SAFE_NAME}.js`);

const copyToPublic = () => exec(`cd dist && npx copyfiles ./* ${DEV_LIVE_MODULES_PATH}`);

(async () => {
  build({
    watch: isProduction
      ? false
      : {
          onRebuild(error) {
            if (!error) {
              console.log('ReBuild succeeded');
              copyToPublic();
            }
          },
        },
    entryPoints: [path.resolve('./src/index.ts')], //entry point
    outbase: 'src',
    bundle: true,
    platform: 'node',
    format: 'cjs', // we should build it as Common JS module
    plugins: [lessLoader({})], // use any plugin you need
    // 1. mark meteor's packages as external
    // 2. some common packages also can be external (ui library etc)
    // 3. our "core" is located in /imports/ folder, ignore it too.
    external: ['mobx', 'react', 'react-dom', 'mobx-react', 'classnames', '@itgenio/*', 'meteor/*', '/imports/*'],
    outfile: OUTPUT_FILE_PATH,
    // banner and footer is for meteor "magic" - meteorInstall
    // generate it like meteor tools do it itself.
    banner: {
      js: `meteorInstall({'node_modules':{${
        isPrefixed ? `'${NAME_PARTS[0]}':{'${NAME_PARTS[1]}'` : '${MODULE_NAME}'
      }(require,exports,module) {`,
    },
    footer: {
      js: `${isPrefixed ? '}' : ''}}}},{extensions:['.js','.jsx','.ts','.tsx','.json','.less','.css','.html','.mjs']})`,
    },
  })
    .then(() => copyToPublic())
    .catch(e => console.error(e.message));
})();
```

In production, we use https://github.com/itgenio/meteor-live-modules (like original dynamic import) or add script to html (like @paulincai  said)

In dev, we copy-paste it to public folder and meteor restarts with new code.

All you need after delivery your package into app - call `require("@company/package1")`

**Pros:**
- the build speed is super fast
- modular/isolated code
- independent release cycle for MC

**Cons:**
- you can't bundle meteor's package into MC (it can be solved with some plugin for esbuild)
- you should manually download/install MC (meteor's dynamic imports "just works")
- isolation of MC brings a new issue with communication between MC(but some good practice live "Event Bus" can solve it)

**Comment by afrokick** [comment]
@renanccastro Another separate step could be replacing `tsc` to `swc`, because tsc works only as a transpiler without type checking.

**Comment by afrokick** [comment]
SWC [will not support nested imports](https://github.com/swc-project/swc/discussions/1800#discussioncomment-3082799), so, it doesn't suit for Meteor as Babel-replacement.

**Comment by StorytellerCZ** [comment]
[Rome](https://rome.tools/) came up a bit in my browsing around the intrawebs. Seems to be quiet interesting project to look into. I think it fits well into Meteor philosophy.

**Comment by boomfly** [comment]
Maybe there is an easy way to add to isobuild multithreading [Nodejs Worker Threads](https://nodejs.org/api/worker_threads.html)?

First build each target in separate thread.

And then compile N files at time. N - CPU cores count.

@paulincai partially right about HMR and almost instant updates in browser during development. BUT when you change server side code, and even more, WHEN you debugging mobile cordova version on each CMD+S I wait 30-60 to see changes on first client code change and on each server side code change.
My dev machine Mac Mini M1

My production build ~30min and 1Gb size
Codebase 75k lines

When I debug cordova native features it's BIG PAIN

**Comment by santiagopuentep** [comment]
We are staying far behind I think.

Vite 3 is out and everything seems to be using it:
https://vitejs.dev/blog/announcing-vite3.html

I think we should ditch isobuild and start using it or a ESM based system or Meteor will never survive long term IMO. The bundler is by far the worst thing about Meteor, especially on Windows. That and not using npm to manage the packages.

**Comment by gunnartorfis** [comment]
Support. Our builds are taking 35 minutes and we've already spent a lot of time trying to reduce that build time. Most of that time is just coming from Meteor build tooling.

**Reply by zodern** [reply]
35 minutes is unusually slow. Could you please try building with the `METEOR_PROFILE=100` env var set, and share the output here?

**Reply by gunnartorfis** [reply]
It actually built in 9 minutes on my M1. Taking 26 minutes and 21 seconds to build the Docker image. We're hosting our Meteor application on Kubernetes. 

Not sure if you'd want to see the Dockerfile?

**Comment by brucejo75** [comment]
Dunno if this may be an issue for your builds, it has been for me...

When I am running a build I make sure I do not have any watchers (`chokidar`, `watcher`, ...) that may be watching the output build directory.  If you have an editor open that the build directory is a subdirectory of will really bog down any file production as the editor watcher will keep on the lookout for new files.

Also, on certain operating systems all new files are automatically scanned for viruses (e.g. windows).  You can configure this scanning to ignore specified directory trees, like your build output.

**Comment by StorytellerCZ** [comment]
Just saw today: https://github.com/Akryum/meteor-vite/blob/main/packages/vite-bundler/README.md

Maybe this might be something to bring into core and replace Babel?

**Reply by afrokick** [reply]
ESBuild doesn't support `emitDecoratorMetadata` option for TypeScript, but enabled in Meteor by default.
Another thing `nested import` and `dynamic import` which work because of `reify` implementation (AFAIK).

**Reply by rj-david** [reply]
Can a plugin like this help: https://github.com/thomaschaaf/esbuild-plugin-tsc ?

For the reify imports, seems something to be raised with akyrum

**Comment by bruceborrett** [comment]
I have been working with Meteor since 2014 and in that time I have also tried many other frameworks, most recently I have worked on a project using Remix and on another project using AWS Amplify, and in my opinion to this day there is still no other framework which offers the same great developer experience as Meteor does. 

However I know that many people do complain about the build times and have stopped using Meteor because of this. For me personally, even with the slow build times it is better than other frameworks out there, but I do think if this one area were improved many people would move back to Meteor or at least want to give it another try.

I am currently working on my first Meteor project in over a year, its an Electron app using meteor-desktop (also a really great project by the way). The initial startup time is around 5 minutes, in the grand scheme of things this is not a deal breaker for me, but my colleagues complain endlessly about this. I can paste my profiling output if it will help, but as others have mentioned above, its probably not worth it to continue debugging isobuild and that time is probably better spent figuring out how to move over to Vite or the like.

Just my 20c.

**Reply by santiagopuentep** [reply]
Are you using Windows or Mac/Linux?

**Reply by bruceborrett** [reply]
Linux (Ubuntu).

**Reply by brucejo75** [reply]
I have tweaked my build process for development to optimize my rebuilds.  Overall it is working really well for me.

In addition to the [tip](https://github.com/meteor/meteor/discussions/11587#discussioncomment-3450285) above...
- to test try a `meteor run` with absolutely nothing else running, then compare to when you have editors open, etc
- on Windows make sure you disable virus scanning for your output folders
- Take a look at all process CPU usage to see if another process spikes while starting up.  It would not be surprising to find other processes that are watching your output folder.  (VSCode really goes nuts if you do not disable watching build folders)
- For most development I only need to build the `modern` build, use the `--exclude-archs` option, e.g. `--exclude-archs=web.browser.legacy,web.cordova`.
- For development you can also `meteor run` with the minimum platforms, e.g. only `web.browser`.
- Make sure you do not perform a `meteor reset` or `rm -r .meteor/local/build` between builds.  Those commands will clear the build cache.  Which can save you a lot of time.

I have used WSL for all development for the last 3-4 years and the build times are very manageable. My machine is a 15 yr old i7 running Windows 10 w/WSL.

**Reply by regardtnel** [reply]
Would love to see what Vite is capable of.

**Comment by italojs** [comment]
awesome work from @nachocodoner adopting SWC to build process
https://github.com/meteor/meteor/pull/13657

**Comment by italojs** [comment]
If you're looking to help speed up our bundler, we've got some great blog posts that will guide you on using our profilers to identify bottlenecks in the building process!

- [Meteor's bundler Profiling](https://dev.to/meteor/unlocking-meteor-32-new-profiling-tool-to-track-bundler-performance-and-size-1jc8)
- [Meteor's cpuProfiling](https://dev.to/italojs/enhancing-performance-in-meteorjs-bundler-complete-guide-to-cpu-profiling-4igo)

**Comment by nachocodoner** [comment]
We’re continuing this conversation in the forums as well: [Join the Effort to Speed Up Meteor bundler](https://forums.meteor.com/t/join-the-effort-to-speed-up-meteor-bundler/63406/10).

There, we share updates on recent changes and performance profiles showing the improvements made so far.

Our efforts to improve build and rebuild times are ongoing! ☄️🚀

**Comment by nachocodoner** [comment]
We delivered already Meteor 3.3 with faster builds/rebuilds x3. See [blog post](https://dev.to/meteor/faster-builds-in-meteor-33-modern-build-stack-with-swc-and-bundler-optimizations-fm2). ☄️🚀

Our next efforts on improve the developer experience is the integration of [Meteor-Rspack](https://forums.meteor.com/t/meteor-rspack-integration-a-modern-bundler-meets-meteor-3-4/63696). Initial testing has showed incredible improvements on speed.

**Comment by KevLehman** [comment]
Can you update your build-cache mechanism so it doesn't grow unbounded?

Frequently I catch `builder-cache` folder on meteor taking more than a couple GBs. Today it took 350GB (as I was doing heavy work on the weekend, changing a bunch of files) and had to spent some time trying to understand why my PC was screaming of no space.

An auto-eviction mechanism as `turbo` implemented on this year could be useful.

**Reply by nachocodoner** [reply]
This is a good idea. Modern bundlers usually have an auto-eviction mechanism, as you mentioned. In the Meteor-Rspack integration, the cache managed by Rspack supports the `maxAge` setting: [https://rspack.rs/config/cache#maxage](https://rspack.rs/config/cache#maxage).

> maxAge (Default: 7 * 24 * 60 * 60 (7 days))
> The maximum time, in seconds, that a cache version can remain unused. During cleanup, **Rspack removes cache versions that have not been accessed for longer than maxAge**. Set maxAge to Infinity to disable age-based cleanup.

However, the `builder-cache` folder you mentioned is managed by Meteor's bundler, so it would be good to have a specific solution on the Meteor side. I could imagine exposing this setting through the `meteor` configuration in `package.json`.

If anyone is willing to provide a solution like this in a PR, I'd be glad to review it and include it in a release. Otherwise, we will keep this suggestion on our list and try to implement it ourselves once we have more time after the current focus and developments.
